javalin/src/main/java/io/javalin/http/staticfiles/PrecompressingResourceHandler.kt

```diff
@@ -0,0 +1,112 @@
+package io.javalin.http.staticfiles
+
+import io.javalin.core.compression.CompressionStrategy
+import io.javalin.core.util.Header
+import io.javalin.http.LeveledBrotliStream
+import io.javalin.http.LeveledGzipStream
+import org.eclipse.jetty.http.MimeTypes
+import org.eclipse.jetty.util.resource.Resource
+import java.io.*
+import javax.servlet.http.HttpServletRequest
+import javax.servlet.http.HttpServletResponse
+
+
+class PrecompressingResourceHandler(val compStrat: CompressionStrategy) {
+
+    val compressedFiles = HashMap<String, ByteArray>()
+
+    companion object {
+        val excludedMimeTypes = setOf(
+                "image/",
+                "audio/",
+                "video/",
+                "application/compress",
+                "application/zip",
+                "application/gzip",
+                "application/bzip2",
+                "application/brotli",
+                "application/x-xz",
+                "application/x-rar-compressed")
+    }
+
+    fun handle(resource: Resource, req: HttpServletRequest, res: HttpServletResponse): Boolean {
+        if (resource.exists() && !resource.isDirectory) {
+            val target = req.getAttribute("jetty-target") as String
+            var acceptCompressType = CompressType.getByAcceptEncoding(req.getHeader(Header.ACCEPT_ENCODING) ?: "")
+            val contentType = MimeTypes.getDefaultMimeByExtension(resource.file.name)//get content type by file extension
+            val resourceCompressible = !excludedMimeType(contentType)
+            if (!resourceCompressible)
+                acceptCompressType = CompressType.NONE
+            val resultByteArray = getStaticResourceByteArray(resource, target, acceptCompressType)
+            res.setContentLength(resultByteArray.size)
+            res.setHeader(Header.CONTENT_TYPE, contentType)
+
+            if (acceptCompressType != CompressType.NONE)
+                res.setHeader(Header.CONTENT_ENCODING, acceptCompressType.typeName)
+
+            val weakETag = resource.weakETag //jetty resource use weakETag too
+            req.getHeader(Header.IF_NONE_MATCH)?.let { etag ->
+                if (etag == weakETag) {
+                    res.status = 304
+                    return true
+                }
+            }
+            res.setHeader(Header.ETAG, weakETag)
+            resultByteArray.inputStream().copyTo(res.outputStream, 2048)
+            res.outputStream.close()
+            return true
+        }
+        return false
+    }
+
+    private fun getStaticResourceByteArray(resource: Resource, target: String, type: CompressType): ByteArray {
+        if (resource.length() > Int.MAX_VALUE)
```

Comment:
Maybe set this to something like 2MB instead, since it's all stored in memory?

Reply:
Sorry. I don't understand what do you mean.
This line is for check that resource size is smaller than 2GB.
Because the maximum size of Resource size is 4GB, but max size of ByteArray in JVM is 2GB.

Reply:
The compressed files are stored in a `Map`, do we want to store 2GB in a `Map`? 

I think it makes more sense to have a much smaller default value (like 2MB), and then `log.warning` if the file is bigger than 2MB? Maybe add a simple config through `PrecompressingResourceHandler.maxSize = ...`.

Reply:
log.warning would be good.
Do we just log.warning then still add the file?
Or log.warning then return or throw exception?

Reply:
log.warning and handle it without compression?

Reply:
> log.warning and handle it without compression?
But it still add to ```Map``` 😕.
How about log.warning then don't use precompression for that static file?

Reply:
It will be
```kotlin
    private fun getStaticResourceByteArray(resource: Resource, target: String, type: CompressType): ByteArray? {
        if (resource.length() > resourceMaxSize) {
            Javalin.log?.warn("PrecompressingResourceHandler can't handle " +
                    "the resource of path : ${resource.file.path} with size : ${resource.length()} byte, " +
                    "because it is larger than resourceMaxSize : $resourceMaxSize byte")
           return null
        }
        val key = target + type.extension
        if (!compressedFiles.containsKey(key)) { // check the target resource exist in compressedFiles or not.
            compressedFiles[key] = getCompressedByteArray(resource, type)
        }
        return compressedFiles[key]!!
    }
```

Reply:
Yes!